### PR TITLE
Episodic metrics should create buffer based on example time step for consistency

### PR DESCRIPTION
**Pull request by breakds**
# Motivation

Currently the episodic metrics class requires a sample metric value to be initialized. However, when a metric value is generated, it is actually extracted from a `TimeStep`. This inconsistency can be harmful as it forces the caller to give different evidence of specification at construction time and usage time of such a metric.

This change is suggested by @hnyu from #1060 

# Solution

The initialization will also use a sample `TimeStep`. The related sub classes are modified accordingly.

# Testing

1. All unit tests passed
2. Using `ac_cart_pole` to verify that the training, evaluation and play are the same. Tensorboard shows the same curve.

tensorboard

### CODE REVIEW

File: alf/metrics/metrics.py [hunk 1]
```diff
@@ -145,33 +146,35 @@ def __init__(self,
...
         """
         super(AverageEpisodicAggregationMetric, self).__init__(
             name=name, dtype=dtype, prefix=prefix)
-        if example_metric_value is None:
+        if example_time_step is None:
             example_metric_value = torch.zeros((), device='cpu')
-        self._batch_size = batch_size
+        else:
+            example_metric_value = self._extract_metric_values(
+                alf.nest.utils.convert_device(example_time_step, device='cpu'))
```
Comment (hnyu) [inline alf/metrics/metrics.py]: Not sure why the original example_metric_value is created on cpu. It's not necessary as I see. Only its structure (not its values) is used.

File: alf/metrics/metrics.py [hunk 2]
```diff
@@ -145,33 +146,35 @@ def __init__(self,
...
+        self._batch_size = alf.nest.get_nest_batch_size(example_time_step)
```
Comment (hnyu) [inline alf/metrics/metrics.py:171]: Do we really need the batch size as a class member?

File: alf/metrics/metrics.py [hunk 3]
```diff
@@ -415,8 +400,9 @@ def __init__(self,
             name=name,
             dtype=dtype,
             prefix=prefix,
-            batch_size=batch_size,
-            buffer_size=buffer_size)
+            buffer_size=buffer_size,
+            example_time_step=TimeStep(
+                step_type=torch.full((batch_size, ), StepType.FIRST)))
```
Comment (hnyu) [inline alf/metrics/metrics.py]: Maybe also pass in the example_time_step instead of creating a dummy one here? Also batch_size will no longer be needed as others. This will unify the constructors of all Metrics.

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/metrics/metrics.py (hunk 1 above)
Comment (breakds) [reply alf/metrics/metrics.py:169]: I think it is because of `AverageDicountedReturnMetric`'s `accumulated_discount`.

File: alf/metrics/metrics.py (hunk 2 above)
Comment (breakds) [reply alf/metrics/metrics.py:171]: It is used 2 times below. I think it does no harm to be kept here.

File: alf/metrics/metrics.py (hunk 1 above)
Comment (hnyu) [reply alf/metrics/metrics.py]: Can directly use `example_time_step.cpu()`

Comment (breakds) [reply alf/metrics/metrics.py]: Fixed.

File: alf/metrics/metrics.py (hunk 3 above)
Comment (breakds) [reply alf/metrics/metrics.py]: Sounds good. Originally I was thinking about trading off consistency for simplicity, but either way is fine.

File: alf/metrics/metrics_test.py [hunk 4]
```diff
@@ -137,16 +132,12 @@ def testMetric(self, metric_class, num_trajectories, expected_result,
...
+                AverageReturnMetric, AverageDiscountedReturnMetric,
+                AverageEnvInfoMetric
         ]:
-
-            metric = metric_class(
-                batch_size=2, reward_shape=(2, ) if vector_reward else ())
-        elif metric_class == AverageEnvInfoMetric:
-            metric = metric_class(
-                batch_size=2, example_env_info=dict(x=0, y=0))
+            metric = metric_class(example_time_step=trajectories[0])
```
Comment (hnyu) [inline alf/metrics/metrics_test.py:138]: Don't need if elif else anymore.

File: alf/metrics/metrics_test.py [hunk 5]
```diff
@@ -166,7 +157,17 @@ def testMetric(self, metric_class, num_trajectories, expected_result,
                              metric.result())
 
     def test_average_per_step(self):
-        metric = AverageDrivingMetric(batch_size=2)
+        metric = AverageDrivingMetric(
```
Comment (hnyu) [inline alf/metrics/metrics_test.py]: Can create this metric after `trajectories` is created, using `trajectories[0]` for metric

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/metrics/metrics_test.py (hunk 5 above)
Comment (breakds) [reply alf/metrics/metrics_test.py]: Good point. Updated.

File: alf/metrics/metrics_test.py (hunk 4 above)
Comment (breakds) [reply alf/metrics/metrics_test.py:138]: Updated with `if .. else ..`. The first two branches are merged.

File: alf/metrics/metrics_test.py [hunk 6]
```diff
@@ -214,6 +212,9 @@ def test_average_per_step(self):
                     },
                     'success': to_tensor([1.0, 0.0])
                 }))
+
+        metric = AverageDrivingMetric(example_time_step=trajectories[0])
```
Comment (hnyu) [inline alf/metrics/metrics_test.py:204]: Should delete the metric in the beginning of this function?

Comment (breakds) [reply alf/metrics/metrics_test.py:204]: Thought I removed it. Sorry.

**Review by hnyu — APPROVED** [review]
